Docker Options Refactor

chadswen opened the pull request:
* Moves `docker_options` out of the `network_plugins` role and into the `docker` role.
* New `all.yml` variable `docker_daemon_graph` for the `--graph` Docker daemon option which was needed across playbooks when changed. I needed this variable for nodes that need Docker to use a specific filesystem.
* The docker role should work now for non-systemd (upstart and sysvinit).
* Modified the conditional checks for CoreOS more accurately mirror systemd vs non-systemd.
* Systemd drop-ins will now be used where systemd is available.
* Systemd and non-systemd docker/flannel options are now configured separately.
* Non-systemd configuration uses lineinfile instead of templates, which allows for idempotency across multiple updates to the same file.
* Proxy support for non-systemd (supersedes PR #575).

Fixes #487 

Updated 11/16/16

bogdando commented on roles/docker/tasks/main.yml:
```diff
@@ -62,6 +62,23 @@
+  when: ansible_os_family != "CoreOS"
+
+- name: Create legacy EnvironmentFile path
+  file: path=/etc/sysconfig state=directory mode=0755
```
Not sure if it's ok to make/change it to worldread

mattymo replied:
/etc/systemd is 0755. this looks fine to me

chadswen replied:
As @mattymo said, that is the default. I could add a when condition to only create the dir when it doesn't exist.

bogdando commented on roles/network_plugin/flannel/templates/docker-network-options-systemd:
```diff
@@ -0,0 +1,2 @@
+[Service]
+Environment="DOCKER_NETWORK_OPTIONS=--bip={{ flannel_subnet }} --mtu={{ flannel_mtu }}"
```
Did I get the change right: that DOCKER_NETWORK_OPTIONS is a uniform option for all OS types, including CoreOS, and it replaces DOCKER_OPTS and OPTIONS variations, which are OS type specific?

chadswen replied:
Yes, the value of `DOCKER_NETWORK_OPTIONS` will be used across all supported OSes. For CoreOS we were using systemd drop-ins exclusively to set options so I continued that here. For the other OSes we were setting `DOCKER_NETWORK_OPTIONS` with `EnvironmentFile`s though they were possibly just copied out of [this example](https://docs.docker.com/engine/admin/systemd/#/custom-docker-daemon-options).

If upstart (Ubuntu 14.04 and 14.10) support is not a concern, I could change this to use drop-ins across all OSes, which would be simpler and much more consistent with recommended systemd usage. Upstart is broken in other ways right now anyway. Thoughts?

chadswen replied:
I thought of yet another way to do this without breaking idempotency or upstart support. Will commit that soon.

mattymo commented:
@chadswen can you rebase this patch?

chadswen commented:
@mattymo Just got back from vacation today. Will do :smiley:

bogdando commented:
@chadswen we have CoreOS CI gates now, could you please rebase to add the change into the CI coverage?

chadswen commented:
Updated. Summary of changes:
* The docker role should work now for non-systemd (upstart and sysvinit).
* Modified the conditional checks for CoreOS more accurately mirror systemd vs non-systemd.
* Systemd drop-ins will now be used where systemd is available.
* Systemd and non-systemd docker/flannel options are now configured separately.
* Non-systemd configuration uses lineinfile instead of templates, which allows for idempotency across multiple updates to the same file.
* Proxy support for non-systemd (supersedes PR #575).

I'm going to run this against my own CI environment scenarios today. Keep the feedback coming :grinning:

bogdando commented:
Thank you! triggered CI as well

chadswen commented:
Made a quick update to surround the docker options values with quotes for non-systemd, which seems to work more consistently with upstart.

chadswen commented:
Strange. Looks like CoreOS failed on one out of three nodes.

@bogdando Have there been changes to the CoreOS test cases that this branch may not have included in that build?

bogdando commented:
@chadswen nothing impacting docker. Try to rebase on the current master as well?

chadswen commented:
@bogdando done. Let's rerun CI.

bogdando commented:
Apiserver fails to come up for some CI cases, we have to investigate

bogdando commented:
It seems we have a fix https://github.com/kubernetes-incubator/kargo/pull/656 , could you please rebase ?

chadswen commented:
@bogdando Rebased.
